theano/gof/vm.py

```diff
@@ -894,6 +935,77 @@ def make_all(self, profiler=None, input_storage=None,
+            node = order[idx]
+            dmap = getattr(node.op, 'destroy_map', None)
+            vmap = getattr(node.op, 'view_map', None)
+            reallocated_ins = set([])
+
+            idx_o = 0
+            for out in node.outputs:
+                ins = None
+                if dmap and idx_o in dmap:
+                    idx_v = dmap[idx_o]
+                    assert len(
+                        idx_v) == 1, "Here we only support the possibility to destroy one input"
+                    ins = node.inputs[idx_v[0]]
+                if vmap and idx_o in vmap:
+                    assert ins is None
+                    idx_v = vmap[idx_o]
+                    assert len(
+                        idx_v) == 1, "Here we only support the possibility to view one input"
+                    ins = node.inputs[idx_v[0]]
+                if ins:
+                    assert isinstance(ins, theano.Variable)
+                    origin = view_of.get(ins, ins)
+                    view_of[out] = origin
+                    viewed_by[origin].append(out)
+                idx_o += 1
+
+            for ins in node.inputs:
+                assert not (ins in view_of and viewed_by[ins])
+                if (getattr(ins, 'ndim', None) == 0 and not storage_map[ins][0] 
+                    and ins not in fgraph.outputs and ins.owner 
+                    and dependencies.get(ins, None)):
+                    # Constant Memory cannot be changed, Constant storage_map
+                    # has a value here
+                    reuse_outs = []
+                    if ins not in view_of and not viewed_by.get(ins, []):
+                        # where gc
+                        for i in range(idx + 1, len(order)):
+                            if reuse_outs:
+                                break
+                            for outs in order[i].outputs:
+                                if outs.ndim == 0 and outs not in viewed_by.get(ins, []):
+                                    reuse_outs.append(storage_map[outs])
+                    elif ins in view_of:
+                        origin = view_of[ins]
+                        viewed_by[origin].remove(ins)
+                        if (not viewed_by[origin] and
+                                origin not in fgraph.inputs and
+                                not isinstance(origin, theano.Constant)):
+                            # where gc
+                            for i in range(idx + 1, len(order)):
+                                if reuse_outs:
+                                    break
+                                for outs in order[i].outputs:
+                                    if outs.ndim == 0 and outs not in viewed_by.get(ins, []):
+                                        reuse_outs.append(storage_map[outs])
+
+                    if reuse_outs:
+                        # if reusable output variable exists
+                        reallocated_ins.add(ins)
```

Comment:
reallocated_ins isn't used. I would get rid of it.

The current code could put in the same storage_map multiple input. You should create a pre_allocated set, where you put all the reuse_outs you got. Also, we only need 1 reuse_out per ins. So keep only one.